src/Data/List/Ext/Subperm.agda

```diff
@@ -0,0 +1,37 @@
+{-# OPTIONS --safe #-}
+
+module Data.List.Ext.Subperm where
+
+open import Prelude
+
+open import Data.List
+open import Data.List.Relation.Unary.Any using (here; there; any?)
+open import Relation.Binary.Definitions using (Decidable)
+open import Data.List.Membership.Propositional using (_∈_)
+
+module _ {A : Set} where
+
+  open import Relation.Binary.Core using (REL)
+
+  data Subperm : REL (List A) (List A) 0ℓ where
+    []   : Subperm [] []
+    _∷ₚʳ_ : ∀ {xs ys} → ∀ y → Subperm xs ys → Subperm xs (y ∷ ys)
+    _∷ₚ_  : ∀ {x xs y ys} → x ∈ (y ∷ ys) → Subperm xs (y ∷ ys) → Subperm (x ∷ xs) (y ∷ ys)
+
+  Subperm⁻¹ : {x : A}{xs l : List A} → Subperm (x ∷ xs) l → x ∈ l × Subperm xs l
+  Subperm⁻¹ (y ∷ₚʳ Spxxs) = (there (proj₁ (Subperm⁻¹ Spxxs))) , (y ∷ₚʳ proj₂ (Subperm⁻¹ Spxxs))
```

Comment:
Instead of repeating the recursive call, I would share via `let`:
```agda
Subperm⁻¹ (y ∷ₚʳ Spxxs) = let x∈, p = Subperm⁻¹ Spxxs in there x∈ , (y ∷ₚʳ p)
```
(unfortunately, Agda will later inline this let and lose this sharing when actually computing with this, but you would at least have written the code in the appropriate way and the fault now lies entirely with the implementation of Agda :P)